text/0625-helper-managers.md

```diff
@@ -0,0 +1,845 @@
+applied for a given render, but before paint. The exact timing may shift around,
+and may or may not correspond to a single rendering pass in cases where there
+are multiple rendering passes in a single paint.
+
+In the future different timings may be added as options for scheduling. For
+instance, a timing to call the effect using `requestIdleCallback`, when the
+browser has finished rendering and handling higher priority work, could be
+added. However, this is out of scope for this RFC.
+
+## How we teach this
+
+Helper managers are a low-level construct that is generally only meant to be
+used by experts and addon authors. As such, it will only be taught through API
+documentation. In addition, for precision and clarity, the API docs will include
+snippets of TypeScript interfaces where appropriate.
+
+### API Docs
+
+#### `setHelperManager`
+
+Sets the helper manager for an object or function.
+
+```js
+setHelperManager((owner) => new ClassHelperManager(owner), Helper.prototype)
+```
+
+When a value is used as a helper in a template, the helper manager is looked up
+on the object by walking up its prototype chain and finding the first helper
+manager. This manager then receives the value and can create and manage an
+instance of a helper from it. This provides a layer of indirection that allows
+users to design high-level helper APIs, without Ember needing to worry about the
+details. High-level APIs can be experimented with and iterated on while the
+core of Ember helpers remains stable, and new APIs can be introduced gradually
+over time to existing code bases.
+
+`setHelperManager` receives two arguments:
+
+1. A factory function, which receives the `owner` and returns an instance of a
+   helper manager.
+2. A helper definition, which is the object or function to associate the factory function with.
+
+The first time the object is looked up, the factory function will be called to
+create the helper manager. It will be cached, and in subsequent lookups the
+cached helper manager will be used instead.
+
+Only one helper manager exists per helper definition, so many helpers will end
+up using the same instance of the helper manager. As such, you should not store
+any state on the helper manager that is related to a single helper instance.
+
+Helper managers must fulfill the following interface (This example uses
+[TypeScript interfaces](https://www.typescriptlang.org/docs/handbook/interfaces.html)
+for precision, you do not need to write helper managers using TypeScript):
+
+```ts
+interface HelperManager<HelperStateBucket> {
+  capabilities: HelperCapabilities;
+
+  createHelper(definition: HelperDefinition, args: TemplateArgs): HelperStateBucket;
+
+  getValue?(bucket: HelperStateBucket, args: TemplateArgs): unknown;
+
+  runEffect?(bucket: HelperStateBucket, args: TemplateArgs): void;
+
+  getDestroyable?(bucket: HelperStateBucket): object;
+}
+```
+
+The capabilities property should be provided using the `capabilities()` function
```

Comment:
It's inconsistent if this is a requirement (*must*) or a recommendation (*should*). It uses *must* just a little bit later.